File: packages/mda-config/typescript/src/trust-policy.ts
import { ErrorCategory, MdaConfigError } from "./errors.js";

export interface SigstoreTrustedSigner {
  type: "sigstore-oidc";
  issuer: string;
  subject: string;
}

export interface DidWebTrustedSigner {
  type: "did-web";
  domain: string;
}

export type TrustedSigner = SigstoreTrustedSigner | DidWebTrustedSigner;

export interface TrustPolicy {
  version: 1;
  trustedSigners: readonly TrustedSigner[];
  minSignatures?: number;
  rekor?: { url: string };
}

/** MDA §13 — validate the trust-policy shape before trusted-runtime use. */
export function validateTrustPolicy(input: unknown): TrustPolicy {
  const policy = requireObject(input, "trustPolicy");
  const allowedTop = new Set(["version", "trustedSigners", "minSignatures", "rekor"]);
  rejectUnknownKeys(policy, allowedTop, "trustPolicy");

  if (policy.version !== 1) {
    policyViolation("trustPolicy.version must be 1");
  }
  if (!Array.isArray(policy.trustedSigners) || policy.trustedSigners.length === 0) {
    policyViolation("trustPolicy.trustedSigners must be a non-empty array");
  }
  const minSignatures = policy.minSignatures;
  if (
    minSignatures !== undefined &&
    (!Number.isInteger(minSignatures) || (minSignatures as number) < 1)
  ) {
    policyViolation("trustPolicy.minSignatures must be an integer >= 1");
  }

  const trustedSigners = policy.trustedSigners.map(validateTrustedSigner);
  const hasSigstore = trustedSigners.some((signer) => signer.type === "sigstore-oidc");
  const hasDidWebOnly = trustedSigners.every((signer) => signer.type === "did-web");
  const rekor = validateRekor(policy.rekor);

  if (hasSigstore && !rekor) {
    policyViolation("Sigstore trust policy entries require rekor.url");
  }
  if (hasDidWebOnly && rekor) {
    policyViolation("did-web-only trust policies must not include rekor");
  }

  return {
    version: 1,
    trustedSigners,
    ...(minSignatures === undefined ? {} : { minSignatures: minSignatures as number }),
    ...(rekor ? { rekor } : {}),
  };
}

export function policyContainsDidWeb(policy: TrustPolicy): boolean {
  return policy.trustedSigners.some((signer) => signer.type === "did-web");
}

export function sigstoreSubjectsFor(policy: TrustPolicy, issuer: string): Set<string> {
  return new Set(
    policy.trustedSigners
      .filter((signer): signer is SigstoreTrustedSigner =>
        signer.type === "sigstore-oidc" && signer.issuer === issuer
      )
      .map((signer) => signer.subject),
  );
}

export function trustsDidWebDomain(policy: TrustPolicy, domain: string): boolean {
  return policy.trustedSigners.some(
    (signer) => signer.type === "did-web" && signer.domain === domain,
  );
}

function validateTrustedSigner(input: unknown): TrustedSigner {
  const signer = requireObject(input, "trustedSigners[]");
  if (signer.type === "sigstore-oidc") {
    rejectUnknownKeys(signer, new Set(["type", "issuer", "subject"]), "sigstore signer");
    if (!isNonEmptyString(signer.issuer) || !isNonEmptyString(signer.subject)) {
      policyViolation("Sigstore trusted signer requires non-empty issuer and subject");
    }
    return { type: "sigstore-oidc", issuer: signer.issuer, subject: signer.subject };
  }
  if (signer.type === "did-web") {
    rejectUnknownKeys(signer, new Set(["type", "domain"]), "did-web signer");
    if (!isNonEmptyString(signer.domain)) {
      policyViolation("did-web trusted signer requires non-empty domain");
    }
    return { type: "did-web", domain: signer.domain };
  }
  policyViolation("trusted signer type must be sigstore-oidc or did-web");
}

function validateRekor(input: unknown): { url: string } | undefined {
  if (input === undefined) return undefined;
  const rekor = requireObject(input, "rekor");
  rejectUnknownKeys(rekor, new Set(["url"]), "rekor");
  if (!isNonEmptyString(rekor.url)) {
    policyViolation("rekor.url must be non-empty");
  }
  return { url: rekor.url };
}

function rejectUnknownKeys(
  object: Record<string, unknown>,
  allowed: Set<string>,
  label: string,
): void {
  for (const key of Object.keys(object)) {
    if (!allowed.has(key)) {
      policyViolation(`${label} has unknown field '${key}'`);
    }
  }
}

function requireObject(input: unknown, label: string): Record<string, unknown> {
  if (!input || typeof input !== "object" || Array.isArray(input)) {
    policyViolation(`${label} must be an object`);
  }
  return input as Record<string, unknown>;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.length > 0;
}

function policyViolation(message: string): never {
  throw new MdaConfigError(ErrorCategory.TrustPolicyViolation, message);
}

File: packages/mda-config/rust/src/trust_policy.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::errors::{ErrorCategory, MdaConfigError, Result};

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TrustPolicy {
    pub version: u8,
    pub trusted_signers: Vec<TrustedSigner>,
    pub min_signatures: Option<usize>,
    pub rekor: Option<RekorPolicy>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RekorPolicy {
    pub url: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
#[serde(tag = "type", rename_all = "kebab-case", deny_unknown_fields)]
pub enum TrustedSigner {
    #[serde(rename = "sigstore-oidc")]
    SigstoreOidc { issuer: String, subject: String },
    #[serde(rename = "did-web")]
    DidWeb { domain: String },
}

impl TrustPolicy {
    pub fn validate(&self) -> Result<()> {
        if self.version != 1 {
            return policy_error("trustPolicy.version must be 1");
        }
        if self.trusted_signers.is_empty() {
            return policy_error("trustPolicy.trustedSigners must be non-empty");
        }
        if self.min_signatures.is_some_and(|value| value == 0) {
            return policy_error("trustPolicy.minSignatures must be >= 1");
        }

        let has_sigstore = self
            .trusted_signers
            .iter()
            .any(|signer| matches!(signer, TrustedSigner::SigstoreOidc { .. }));
        let has_did_web = self
            .trusted_signers
            .iter()
            .any(|signer| matches!(signer, TrustedSigner::DidWeb { .. }));

        for signer in &self.trusted_signers {
            match signer {
                TrustedSigner::SigstoreOidc { issuer, subject } => {
                    if issuer.is_empty() || subject.is_empty() {
                        return policy_error(
                            "Sigstore trust policy entries require issuer and subject",
                        );
                    }
                }
                TrustedSigner::DidWeb { domain } => {
                    if domain.is_empty() {
                        return policy_error("did:web trust policy entries require domain");
                    }
                }
            }
        }

        if has_sigstore && self.rekor.as_ref().is_none_or(|rekor| rekor.url.is_empty()) {
            return policy_error("Sigstore trust policy entries require rekor.url");
        }
        if has_did_web && !has_sigstore && self.rekor.is_some() {
            return policy_error("did:web-only trust policy must not include rekor");
        }
        Ok(())
    }

    pub fn required_signatures(&self) -> usize {
        self.min_signatures.unwrap_or(1)
    }

    pub fn contains_did_web(&self) -> bool {
        self.trusted_signers
            .iter()
            .any(|signer| matches!(signer, TrustedSigner::DidWeb { .. }))
    }

    pub fn sigstore_subject_trusted(&self, issuer: &str, subject: &str) -> bool {
        self.trusted_signers.iter().any(|signer| {
            matches!(
                signer,
                TrustedSigner::SigstoreOidc {
                    issuer: trusted_issuer,
                    subject: trusted_subject
                } if trusted_issuer == issuer && trusted_subject == subject
            )
        })
    }

    pub fn sigstore_issuer_trusted(&self, issuer: &str) -> bool {
        self.trusted_signers.iter().any(|signer| {
            matches!(
                signer,
                TrustedSigner::SigstoreOidc {
                    issuer: trusted_issuer,
                    ..
                } if trusted_issuer == issuer
            )
        })
    }

    pub fn did_web_domain_trusted(&self, domain: &str) -> bool {
        self.trusted_signers.iter().any(|signer| {
            matches!(
                signer,
                TrustedSigner::DidWeb {
                    domain: trusted_domain
                } if trusted_domain == domain
            )
        })
    }
}

pub fn validate_trust_policy(value: Value) -> Result<TrustPolicy> {
    let policy: TrustPolicy = serde_json::from_value(value).map_err(|cause| {
        MdaConfigError::with_details(
            ErrorCategory::TrustPolicyViolation,
            "trust policy does not match RC2 schema",
            serde_json::json!({ "cause": cause.to_string() }),
        )
    })?;
    policy.validate()?;
    Ok(policy)
}

fn policy_error<T>(message: &str) -> Result<T> {
    Err(MdaConfigError::new(
        ErrorCategory::TrustPolicyViolation,
        message,
    ))
}
